ctapipe/io/hessiofilereader.py

```diff
@@ -0,0 +1,211 @@
+from astropy import units as u
+from astropy.coordinates import Angle
+from astropy.time import Time
+from ctapipe.core import Provenance
+from ctapipe.io.eventfilereader import EventFileReader
+from ctapipe.io.containers import DataContainer
+from ctapipe.instrument import TelescopeDescription, SubarrayDescription
+
+
+COUNT = 0
+
+
+class HessioFileReader(EventFileReader):
+
+    def __init__(self, config, tool, **kwargs):
+        super().__init__(config=config, tool=tool, **kwargs)
+
+        try:
+            from pyhessio import open_hessio
+            from pyhessio import close_file
+            from pyhessio import HessioError
+            from pyhessio import HessioTelescopeIndexError
+            from pyhessio import HessioGeneralError
+        except ImportError:
+            msg = "The `pyhessio` python module is required to access MC data"
+            self.log.error(msg)
+            raise
+
+        self.open_hessio = open_hessio
+        self.close_hessio = close_file
+        self.HessioError = HessioError
+        self.HessioTelescopeIndexError = HessioTelescopeIndexError
+        self.HessioGeneralError = HessioGeneralError
+
+        self.allowed_tels = None
+
+        global COUNT
+        if COUNT > 0:
+            self.log.warn("Only one pyhessio reader allowed at a time. "
+                          "Previous hessio file will be closed.")
+            self.close_hessio()
+        COUNT += 1
+
+    @staticmethod
+    def is_compatible(file_path):
+        return file_path.endswith('.gz')
+
+    @property
+    def camera(self):
```

Comment:
Why is it called `camera`? At least HessIO files contain Array Events with different cameras.

Reply:
In the future there will be only one camera per file.

This exists so that scripts decide at run time the correct default operations to use according to the camera, for example the correct CameraR1Calibrator or ChargeExtractor. In my next PR I will create a class that takes a camera name and sets up the traitlet config so the correct Components are used for the camera.

It makes sense for the camera field to exist here as the camera type can be supplied from the file (either from a header or inferred from the data format).

Hessio files require the HessioR1Calibrator to be applied to them. So that will be set as the default R1Calibrator when given "hessio" as the input. This interface will be further improved in the future when we do have single cameras per file, and then camera-specific operations can be defined for simulation files.

Reply:
perhaps this is meant to give the file type? In which case it could be renamed? I'm not sure why it is there either.

Reply:
@kosack If the `inst` container could somehow indicate that the file is a "simulated" camera from sim_telarray, and therefore HessioR1Calibrator should be used, then we could remove the `EventFileReader.camera` property, and rely on the `inst` container for such information.

Reply:
Yes, I think we can add that functionality.  I can just add an "is_simulated" attribute to SubarrayDescription.  

FileReaders should only fill in the cameras that are in the file into SubarrayDescription, so then you can query it to see which camera is there (and we support both merged and unmerged camera files simultaneously).  I may have to rethink the naming, or add another description class, since "Subarray" should really refer to all the tels, not just the one in the file.

I have another suggestion: since all input files may have some generic headers (version number, etc) defined, why not just have a method or attribute called "metadata" that returns for now a dict, and you can keep an item like CAMERA in that? That way we can pyt other into in it, like version since the schema of the format may evolve.  In fact, it should really be "TELESCOPE" not "CAMERA" since it depends also on the OpticsDescription, etc. So it should be filled with a value that is the same as  what `str(TelescopeDescription)` gives you (like "SST-GCT:CHEC")

Reply:
So more concretely, for now just have:

```py3
def metadata(self):
     return self.metadata
```
where `self.metadata` is filled in when the file is opened by the particular subclass, and can minimally be just `{'telescope': tel_identifier}`, and later we can also add things like the `{'obs_id' : 12345, is_simulation:True}`, etc. (to be defined)

Reply:
or you could write that info into `event.meta` which is already there. We just have to define a common place to put file metadata.

Reply:
Maybe `event.meta` isn't the best place, as you would need to retrieve an event container (by beginning to loop through the events) to be able to retrieve that information... And also it is information that does not change with event.

I prefer the `reader.metadata` and `reader.subarray` attributes.

Reply:
Should we create a metadata class instead of using a dictionary so the contents are consistent and predictable?